Source/DafnyTestGeneration/README.md

```diff
@@ -0,0 +1,149 @@
+
+While these manipulations could be done with the Dafny source directly, we instead modify the Boogie translation, since it is much easier to work with basic blocks and paths on this lower level.
+
+Objects present an obstacle for test generation since it is not possible to mock objects in Dafny and the prover model does not specify how an object was allocated\constructed. Currently, we require that a test method take any object it needs as an argument. This allows us to compile the test methods to Java or another language where object mocking is possible (see [How to Run Tests?](#how-to-run-tests)).
+
+Even though it is possible to run the tests in this way, we currently provide no oracle against which to compare the result. In the future, it may be possible to generate an oracle directly from the prover model provided that there is no non-determinism in the tested code.
+
+
+## Implementation
+
+- [Main.cs](Main.cs) - contains the two methods that should be used to query Dafny to generate tests.
+- [Utils.cs](Utils.cs) - a small collection of utility methods used elsewhere in the project.
+- [TestMethod.cs](TestMethod.cs) - generates a test that calls the target method with arguments triggering an assertion violation described by Z3's counterexample model. This is where all the interaction with DafnyServer's counterexample extraction code takes place.
+- [ProgramModification.cs](ProgramModification.cs) - represents a modification of the original Boogie program (which is in turn generated in `Main.cs` by translating Dafny input to Boogie). Can be verified and, if verification fails, one can pass the counterexample log as constructor argument to `TestMethod`.
+- [ProgramModifier.cs](ProgramModifier.cs) - an abstract class that contains utilities for annotating Boogie Programs, inlining Dafny methods after they are translated to Boogie, etc. This is all necessary for generating instances of `ProgramModification`
+- [BlockBasedModifier.cs](BlockBasedModifier.cs) and [PathBasedModifier.cs](PathBasedModifier.cs) - subclasses of `ProgramModifier` that generate instances of `ProgramModification` by inserting assertions that fail when a particular block is visited or a particular path is taken, respectively.
+- [DafnyInfo.cs](DafnyInfo.cs) - takes a Dafny Program and extracts certain information about it that is then used in `TestMethod.cs` and `Main.cs`
+
+## How to Generate Tests?
+
+- Test generation currently works with all basic types, user-defined classes, sequences, sets, and maps. It does not work with datatypes, arrays, and multisets. It is also not possible to generate tests for constructors.
+- To generate block- or path-coverage tests use the `/testMode:Block` or `/testMode:Path` arguments respectively. You will likely also need `definiteAssignment:3`, which will ensure that you generate tests even for those methods that do not contain any assertions or pre-\post-conditions.
```

Comment:
Let's use `generateTest...` instead of just `/test...` as the common prefix for these options, to avoid implying that we're actually testing or running tests. If and when we separate out the different `dafny` commands as suggested in https://github.com/dafny-lang/dafny/issues/1322#issuecomment-889023432, we can make this feature a separate `dafny generate-tests ...` command.

Reply:
Have made the change in one of my latest commits. Separating this into a standalone `generate-tests` command would be great for reducing argument name length.